rfd/0078-role-variables.md

```diff
@@ -0,0 +1,176 @@
+
+## Why
+
+Variables will be useful to specify additional traits that are not available in
+OIDC/SAML claims.
+
+Teleport admins don't always have control over their org's OIDC/SAML provider.
+
+Role variables will simplify role evaluation, keep logic in one place, and
+minimize repetition.
+
+## Details
+
+Example role spec:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: node-users
+spec:
+  vars:
+    - name: "logins"
+      # "select" is a string value which will be used as the input to all
+      # possible expansion matchers. It can (and usually will) contain a trait
+      # which will be expanded before passing as input.
+      select: "{{external.email}}"
+      expansions:
+        # nic should be able to login as root
+        - match: '^nic@goteleport\.com$'
+          out: [root]
+        # alice should be able to login as root and admin
+        - match: '^alice@goteleport\.com$'
+          out: [root, admin]
+        # all teleport employees should be able to login as the local part of their email
+        - match: '@goteleport\.com$'
+          # "out" values can contain transforms and any traits
+          out: ["{{email.local(external.email)}}"]
+    - name: "allow_env"
+      # If the trait in a "select" expands to a list, all values will be checked
+      # for matches and the outputs will be combined.
+      select: "{{external.groups}}"
+      expansions:
+        # matchers can contain regexp capture groups, which can be expanded in
+        # "out" values
+        - match: '^env-(\w+)$'
+          out: ['$1']
+        - match: '^admins$'
+          out: ["prod"]
+  allow:
+    logins:
+      - "{{vars.logins}}"
+    node_labels:
+      env: ["{{vars.allow_env}}"]
+```
+
+These variables could be used in another role without redefinition:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: app-users
+spec:
+  allow:
+    app_labels:
+      env: ["{{vars.allow_env}}"]
+```
+
+### Should we select the first matching expansion, or all matching expansions?
+
+For the example
+
+```yaml
+vars:
+  - name: "logins"
+    select: "{{external.username}}"
+    expansions:
+      - match: "nic"
+        out: [root]
+      - match: "bob"
+        out: [readonly]
+      - match: "(.*)"
+        out: ["$1"]
+```
+
+we could either choose to expand the first match or all matches, the results for
+which are shown in this table:
+
+username | first match | all matches
+---------|-------------|------------
+nic      | [root]      | [root, nic]
+bob      | []          | [bob]
+alice    | [alice]     | [alice]
+
+In favor of expanding all matches, it is useful to be able to provide a default
+which applies to all possible values.
+
+In favor of selecting only the first match, it is useful to be able to exclude
+some value from matching any other expansions.
+This is especially true given Go's lack of support for negative-lookahead
+regular expressions.
+
+I'm not quite sure which is the best route to take here and am quite open to
```

Comment:
I suppose I could also just add a `not_match: <regex>` field to allow negative matches.

Reply:
I have updated the syntax so that each possible value can have its own unique input, with a `match` and `not_match` expression, and this question is no longer relevant.